### PR TITLE
Updated code review assignment logic.

### PR DESCRIPTION
**Pull request by russjones**
**Description**

Code reviewers are split into code owners and code reviewers. Each PR is now randomly assigned a reviewers from the code owner and reviewer set. Review checking makes sure that a reviewer from each set has approved the PR.

### CODE REVIEW

File: .github/workflows/robot/internal/review/review.go
Comment (quinqu) [inline .github/workflows/robot/internal/review/review.go:228]:
```suggestion
	if check(r.getAdminReviewers(author), reviews) {
```

**Review by quinqu — COMMENTED** [review]
I will do another round in a bit, but have a few qs and suggestions for now.

File: .github/workflows/assign.yaml
Comment (quinqu) [inline .github/workflows/assign.yaml:11]: I don't think this workflow will assign additional reviewers if the file types change (ex: PR opens with no docs changes then author pushes a commit with docs changes).

Comment (quinqu) [reply .github/workflows/assign.yaml:11]:
```suggestion
    types: [assigned, opened, reopened, ready_for_review, synchronize]
```

File: .github/workflows/assign.yaml
Comment (tcsc) [inline .github/workflows/assign.yaml:44]: I've been thinking about this, and perhaps we're thinking about the bot in the wrong way. We've been thinking about **_The_** bot, singular, when in reality there are multiple, semi-independent tasks that need to happen that could be structured as independent scripts. They'd all share some back-end code, yes, but splitting them out could let us simplify each script down to its essentials without having to worry as much about compatibility with the other tasks.

I've not sat down and thought on this deeply yet, so it may be a terrible idea!

File: .github/workflows/robot/internal/review/review.go
Comment (russjones) [reply .github/workflows/robot/internal/review/review.go:156]: Check requires a code owner and code reviewer approval.

File: .github/workflows/robot/internal/review/review.go
Comment (russjones) [reply .github/workflows/robot/internal/review/review.go]: Updated, now internal assignment still goes to default reviews, but internal check requires same code owner and reviewer.

File: .github/workflows/robot/internal/review/review.go
Comment (tcsc) [inline .github/workflows/robot/internal/review/review.go:135]: ? 

```suggestion
		reviewers = append(reviewers, r.getAdminReviewers(author)...)
```

Based on comment above

File: .github/workflows/robot/internal/review/review.go
Comment (tcsc) [inline .github/workflows/robot/internal/review/review.go:142]: Magic team name. It would be nice to have these extracted to constants with accompanying docs saying what they each team name implies.

File: .github/workflows/robot/internal/review/review.go
Comment (tcsc) [inline .github/workflows/robot/internal/review/review.go:176]: Another magic team name

File: .github/workflows/robot/internal/review/review.go
Comment (tcsc) [inline .github/workflows/robot/internal/review/review.go:239]:
```suggestion
func (r *Assignments) checkCodeReviews(authorName string, reviews map[string]*github.Review) error {
```

File: .github/workflows/robot/internal/review/review.go
Comment (tcsc) [inline .github/workflows/robot/internal/review/review.go:242]:
```suggestion
	author, ok := r.c.CodeReviewers[authorName]
```

File: .github/workflows/robot/internal/review/review.go
Comment (tcsc) [inline .github/workflows/robot/internal/review/review.go:249]: Magic team names.

File: .github/workflows/robot/internal/review/review.go
Comment (quinqu) [reply .github/workflows/robot/internal/review/review.go:135]: +1

**Review by quinqu — APPROVED** [review]

File: .github/workflows/robot/internal/review/review.go
Comment (russjones) [reply .github/workflows/robot/internal/review/review.go:142]: I understand, however I don't think it makes sense to make these constants+docs at the moment.

It's confusion because I had to pull out the reviewers into a secret, once this moves to a private GitHub repo and is converted into an action we can bring that back in and we can make the change then.

File: .github/workflows/assign.yaml
Comment (russjones) [reply .github/workflows/assign.yaml:7]: Not sure I understand, I thought the token had full access to GitHub?

Comment (russjones) [reply .github/workflows/assign.yaml]: Synced with Trent on this, updated the wording of this comment.

**Review by tcsc — APPROVED** [review]

**Review by r0mant — APPROVED** [review]

File: .github/workflows/robot/internal/bot/dismiss.go
Comment (quinqu) [inline .github/workflows/robot/internal/bot/dismiss.go:47]:
```suggestion
		if err := b.dismiss(ctx, pull.Organization, pull.Repository, pull.UnsafeHead); err != nil {
```

**Pull request merged**
